src/HealthGPS/analysis_module.cpp

```diff
@@ -26,6 +26,46 @@ SimulationModuleType AnalysisModule::type() const noexcept {
     return SimulationModuleType::Analysis;
 }
 
+void AnalysisModule::initialise_map(RuntimeContext &context) {
+    // Set some arbitrary factors
+    auto factor1 = context.mapping().entries().at(0).key().to_string();
+    auto factor2 = context.mapping().entries().at(1).key().to_string();
+
+    // Get the minimum and maximum of the two factors in the population
+    auto min_max = std::minmax_element(context.population().cbegin(), context.population().cend(),
+                                       [](const auto &a, const auto &b) {
+                                           return a.get_risk_factor_value(factor1) <
+                                                  b.get_risk_factor_value(factor1);
+                                       });
+
+    auto min_factor1 = min_max.first->get_risk_factor_value(factor1);
+    auto max_factor1 = min_max.second->get_risk_factor_value(factor1);
+
+    min_max = std::minmax_element(context.population().cbegin(), context.population().cend(),
+                                  [](const auto &a, const auto &b) {
+                                      return a.get_risk_factor_value(factor2) <
+                                             b.get_risk_factor_value(factor2);
+                                  });
+
+    auto min_factor2 = min_max.first->get_risk_factor_value(factor2);
+    auto max_factor2 = min_max.second->get_risk_factor_value(factor2);
+
+    // The dimensions of the 2D map is the number of integer values of each factor, or 100 bins of
+    // equal size, whichever is smaller (100 is an arbitrary number, it could be any other number
+    // depending on the desired resolution of the map)
+    auto factor1_bins = std::min(100, static_cast<int>(max_factor1 - min_factor1));
+    auto factor2_bins = std::min(100, static_cast<int>(max_factor2 - min_factor2));
+
+    // Create a 2D map with the specified number of bins
+    calculated_factors_ =
+        std::unordered_map<int, std::unordered_map<int, std::unordered_map<std::string, double>>>{};
```

Comment:
I'm guessing the keys will be in some range 0 to `n` for the outer map and 0 to `m` for the inner one? If that's the case, then you don't need maps; you could just use `std::vector`s instead, which will be faster.

I.e. the type would be `std::vector<std::vector<std::unordered_map<core::Identifier, double>>>`

Reply:
(Technically you could just use a single `std::vector` and then access the right element with `i * row_size + j` which will be faster still, but I don't know whether we care that much about performance here. It's also *potentially* a bit cleaner, because it makes it easier to do things to the whole data structure at once.)